src/upsert.rs

```diff
@@ -34,59 +32,81 @@ pub trait FindexUpsert<
         master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
         label: &Label,
-        additions: HashMap<IndexedValue, HashSet<Keyword>>,
-        deletions: HashMap<IndexedValue, HashSet<Keyword>>,
+        items: HashMap<IndexedValue, HashSet<Keyword>>,
     ) -> Result<(), Error<CustomError>> {
-        check_parameter_constraints::<CHAIN_TABLE_WIDTH, BLOCK_LENGTH>();
-
-        let mut rng = CsRng::from_entropy();
-        let k_uid: KmacKey = master_key.derive_kmac_key(ENTRY_TABLE_KEY_DERIVATION_INFO);
-        let k_value = master_key.derive_dem_key(ENTRY_TABLE_KEY_DERIVATION_INFO);
-
-        // Revert the `HashMap`.
-        let mut new_chains = HashMap::<KeywordHash, HashMap<IndexedValue, BlockType>>::default();
-        for (indexed_value, keywords) in additions {
+        let mut new_chains = HashMap::<Keyword, HashMap<IndexedValue, BlockType>>::default();
+        for (indexed_value, keywords) in items {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Addition);
             }
         }
-        // Adding deletions after additions only keeps deletions.
-        for (indexed_value, keywords) in deletions {
+        self.upsert(master_key, label, new_chains).await
+    }
+
+    /// Removes the given values from the indexes for the associated keywords.
+    ///
+    /// # Parameters
+    ///
+    /// - `master_key`  : Findex master key
+    /// - `label`       : additional public information used in key hashing
+    /// - `items`       : set of keywords used to index values
+    async fn remove(
+        &mut self,
+        master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
+        label: &Label,
+        items: HashMap<IndexedValue, HashSet<Keyword>>,
+    ) -> Result<(), Error<CustomError>> {
+        let mut new_chains = HashMap::<Keyword, HashMap<IndexedValue, BlockType>>::default();
+        for (indexed_value, keywords) in items.into_iter() {
             for keyword in keywords {
                 new_chains
-                    .entry(keyword.hash())
+                    .entry(keyword)
                     .or_default()
                     .insert(indexed_value.clone(), BlockType::Deletion);
             }
         }
+        self.upsert(master_key, label, new_chains).await
+    }
+
+    /// Upsert the given chain elements in Findex tables.
+    ///
+    /// # Parameters
+    ///
+    /// - `master_key`      : Findex master key
+    /// - `label`           : additional public information used in key hashing
+    /// - `modifications`   : set of operations to apply on the associated value
+    ///   for the associated keyword
+    async fn upsert(
+        &mut self,
+        master_key: &KeyingMaterial<MASTER_KEY_LENGTH>,
+        label: &Label,
+        modifications: HashMap<Keyword, HashMap<IndexedValue, BlockType>>,
```

Comment:
I'm not sure I like the exposition of `BlockType` to the client. I may prefer the old way with two parameters… What do you think @Manuthor?

Reply:
I suppose it has been made to avoid an allocation. IMHO, additions and deletions are more readable

Reply:
I was thinking of it as a way to remove this ugly piece of code from the `upsert`:
```        let mut new_chains =
            HashMap::<KeywordHash, HashMap<IndexedValue, BlockType>>::with_capacity(
                additions.len() + deletions.len(),
            );
        for (indexed_value, keywords) in additions {
            for keyword in keywords {
                new_chains
                    .entry(keyword.hash())
                    .or_default()
                    .insert(indexed_value.clone(), BlockType::Addition);
            }
        }
        for (indexed_value, keywords) in deletions {
            for keyword in keywords {
                new_chains
                    .entry(keyword.hash())
                    .or_default()
                    .insert(indexed_value.clone(), BlockType::Deletion);
            }
        }
```

Reply:
But I removed in it the last commits.